extensions/spatial/src/org/sbml/jsbml/validator/offline/constraints/GeometryConstraints.java

```diff
@@ -0,0 +1,410 @@
+	   */
+	  @Override
+	  public void addErrorCodesForCheck(Set<Integer> set, int level, int version,
+	    CHECK_CATEGORY category, ValidationContext context) {
+		  switch (category) {
+		    case GENERAL_CONSISTENCY:
+		    	if(level >= 3){		    		
+		    		addRangeToSet(set, SPATIAL_23701, SPATIAL_23717);
+		    	}
+		      break;
+		    case IDENTIFIER_CONSISTENCY:
+		      break;
+		    case MATHML_CONSISTENCY:
+		      break;
+		    case MODELING_PRACTICE:
+		      break;
+		    case OVERDETERMINED_MODEL:
+		      break;
+		    case SBO_CONSISTENCY:
+		      break;
+		    case UNITS_CONSISTENCY:
+		      break;
+		    }
+	  }
+	  
+	  @Override
+	  public ValidationFunction<?> getValidationFunction(int errorCode, ValidationContext context){
+		  ValidationFunction<Geometry> func = null;
+		  
+		  switch (errorCode) {
+		  
+		  	case SPATIAL_23701:
+		  	{
+		  		// A Geometry object may have the optional SBML Level 3 Core attributes metaid and sboTerm. 
+		  		// No other attributes from the SBML Level 3 Core namespaces are permitted on a Geometry.
+		  		
+		  		func = new UnknownCoreAttributeValidationFunction<Geometry>();
+		  		break;
+		  	}
+		  	
+		  	case SPATIAL_23702:
+		  	{
+		  		// A Geometry object may have the optional SBML Level 3 Core subobjects for notes and annotations.
+		  		// No other elements from the SBML Level 3 Core namespaces are permitted on a Geometry.
+		  		
+		  		func = new UnknownCoreElementValidationFunction<Geometry>();
+		  		break;
+		  	}
+		  	
+		  	case SPATIAL_23703:
+		  	{
+		  		// A Geometry object must have the required attribute spatial:coordinateSystem, and may 
+		  		// have the optional attribute spatial:id. No other attributes from the SBML Level 3 Spatial
+		  		// Processes namespaces are permitted on a Geometry object.
+		  		
+		  		func = new UnknownPackageAttributeValidationFunction<Geometry>(SpatialConstants.shortLabel) {
+		  			
+		  			@Override
+		  			public boolean check(ValidationContext ctx, Geometry geom) {
+		  				
+		  				if(!geom.isSetCoordinateSystem()) {
+		  					return false;
+		  				}
+		  				
+		  				return super.check(ctx, geom);
+		  			}
+		  		};
+		  		break;
+		  	}
+		  	
+		  	case SPATIAL_23704:
+		  	{
+		  		// A Geometry object may contain one and only one instance of each of the ListOfCoordinateComponents,
+		  		// ListOfDomainTypes, ListOfDomains, ListOfAdjacentDomains, ListOfGeometryDefinitions and
+		  		// ListOfSampledFields elements. No other elements from the SBML Level 3 Spatial Processes
+		  		// namespaces are permitted on a Geometry object.
+		  		
+		  		func = new ValidationFunction<Geometry>() {
+		  			
+		  			@Override
+		  			public boolean check(ValidationContext ctx, Geometry geom) {
+		  				
+		  				boolean locc = geom.isSetListOfCoordinateComponents() 
```

Comment:
You still need to have a `if (isSetListOf)` for each test because otherwise it does create an empty list that you don't want to create and go through a method call that is not necessary.

If the specific rule 23704 does not ask you to test if listOfCoordinateComponents is present, just if it is not duplicated, you should not test more. Don't fail a check for reason that are not part of the rule :-)

Reply:
You could have one boolean only defined to true at the start, then you use it for each if statement:
`
boolean check = true;
if (check && isSetListof)  {
check &= "duplicated test"
}
...

return check;
`

Reply:
Oh, right! I understand it now! :-)
That check for listOfCoordinateComponents can be added later as a rule added by hand.
We are checking for duplicates only if that ListOf<> element is set, right?

Reply:
yes